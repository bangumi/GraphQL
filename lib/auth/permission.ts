import * as php from 'php-serialize';
import NodeCache from 'node-cache';

import prisma from '../prisma';
import { logger } from '../logger';

const cache = new NodeCache({ stdTTL: 60 * 10 });

export interface Permission {
  user_list?: boolean;
  manage_user_group?: boolean;
  manage_user_photo?: boolean;
  manage_topic_state?: boolean;
  manage_report?: boolean;
  user_ban?: boolean;
  manage_user?: boolean;
  user_group?: boolean;
  user_wiki_approve?: boolean;
  doujin_subject_erase?: boolean;
  user_wiki_apply?: boolean;
  doujin_subject_lock?: boolean;
  subject_edit?: boolean;
  subject_lock?: boolean;
  subject_refresh?: boolean;
  subject_related?: boolean;
  subject_merge?: boolean;
  subject_erase?: boolean;
  subject_cover_lock?: boolean;
  subject_cover_erase?: boolean;
  mono_edit?: boolean;
  mono_lock?: boolean;
  mono_merge?: boolean;
  mono_erase?: boolean;
  ep_edit?: boolean;
  ep_move?: boolean;
  ep_merge?: boolean;
  ep_lock?: boolean;
  ep_erase?: boolean;
  report?: boolean;
  manage_app?: boolean;
  app_erase?: boolean;
}

export async function getPermission(userGroup: number): Promise<Readonly<Permission>> {
  const cached = cache.get(userGroup);
  if (cached) {
    return cached;
  }

  const permission = await prisma.chii_usergroup.findFirst({ where: { usr_grp_id: userGroup } });
  if (!permission) {
    logger.warn("can't find permission for userGroup %d", userGroup);
    return {};
  }

  const p = Object.freeze(
    Object.fromEntries(
      Object.entries(
        php.unserialize(permission.usr_grp_perm) as Record<keyof Permission, string>,
      ).map(([key, value]) => [key, value === '1']),
    ),
  );

  cache.set(userGroup, p);

  return p;
}
