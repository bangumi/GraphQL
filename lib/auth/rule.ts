import type { IReply } from '../orm';
import { ReplyState } from '../topic';
import type { IAuth } from './index';

export const enum TopicDisplay {
  Ban = 0,
  Normal = 1,
  Review = 2,
}

export const rule = {
  ListTopicDisplays(u: IAuth): TopicDisplay[] {
    if (!u.login) {
      return [TopicDisplay.Normal];
    }

    if (u.permission.manage_topic_state) {
      return [TopicDisplay.Ban, TopicDisplay.Normal, TopicDisplay.Review];
    }

    return [TopicDisplay.Normal];
  },

  filterReply(x: IReply): IReply {
    return { ...filterReply(x), replies: x.replies.map((x) => filterReply(x)) };
  },
};

function filterReply<T extends { state: number; text: string }>(x: T): T {
  if (
    x.state === ReplyState.AdminDelete ||
    x.state === ReplyState.UserDelete ||
    x.state === ReplyState.AdminReopen ||
    x.state === ReplyState.AdminCloseTopic ||
    x.state === ReplyState.AdminSilentTopic
  ) {
    return { ...x, text: '' };
  }

  return x;
}
