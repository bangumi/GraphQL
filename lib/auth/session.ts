import prisma from '../prisma';
import { randomBase62String } from '../utils';
import * as auth from './index';
import type { IAuth } from './index';

export async function create(user: { id: number; regTime: number }): Promise<string> {
  const now = Math.trunc(Date.now() / 1000);
  const token = randomBase62String(32);
  const value = {
    reg_time: new Date(user.regTime * 1000).toISOString(),
    user_id: user.id,
    created_at: now,
    expired_at: now + 60 * 60 * 24 * 7,
  };

  await prisma.chii_os_web_sessions.create({
    data: {
      value: Buffer.from(JSON.stringify(value)),
      user_id: user.id,
      created_at: value.created_at,
      expired_at: value.expired_at,
      key: token,
    },
  });

  return token;
}

/**
 * TODO: add cache
 * @param sessionID store in user cookies
 */
export async function get(sessionID: string): Promise<IAuth | null> {
  const session = await prisma.chii_os_web_sessions.findFirst({
    where: { key: sessionID, expired_at: { gte: Math.trunc(Date.now() / 1000) } },
  });
  if (!session) {
    return null;
  }

  return await auth.byUserID(session.user_id);
}

export async function revoke(sessionID: string) {
  await Promise.all([
    prisma.chii_os_web_sessions.update({
      where: {
        key: sessionID,
      },
      data: {
        expired_at: Math.trunc(Date.now() / 1000) - 60 * 60,
      },
    }),
  ]);
}
