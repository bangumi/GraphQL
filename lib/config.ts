import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import url from 'node:url';

import type { RedisOptions } from 'ioredis';

export const production = process.env.NODE_ENV === 'production';
export const testing = process.env.NODE_ENV === 'test';

export const projectRoot = url.fileURLToPath(new URL('..', import.meta.url));
export const pkg = JSON.parse(
  fs.readFileSync(path.resolve(projectRoot, 'package.json'), 'utf8'),
) as { version: string };

export const redisPrefix = `graphql-${pkg.version}`;

const u = url.parse(process.env.REDIS_URI ?? 'redis://127.0.0.1:3306/0');

const [username, password] = (u.auth ?? '').split(':', 2);

export const redisOption = {
  host: u.hostname ?? '127.0.0.1',
  port: u.port ? Number.parseInt(u.port) : 3306,
  db: u.pathname ? Number.parseInt(u.pathname.slice(1)) : 0,
  username: username,
  password: password,
  lazyConnect: true,
} satisfies RedisOptions;
