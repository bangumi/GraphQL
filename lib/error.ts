import type { FastifyError } from '@fastify/error';
import { createError } from '@fastify/error';
import httpCodes, { StatusCodes } from 'http-status-codes';

/** Type helper to make message argument as required */
export interface SingleMessageErrorConstructor {
  new (msg: string): FastifyError & { statusCode: number };
  (msg: string): FastifyError & { statusCode: number };
}

export const BadRequestError: SingleMessageErrorConstructor = createError(
  'BAD_REQUEST',
  '%s',
  StatusCodes.BAD_REQUEST,
);
export const NotFoundError: SingleMessageErrorConstructor = createError(
  'NOT_FOUND',
  '%s not found',
  httpCodes.NOT_FOUND,
);
export const UnexpectedNotFoundError: SingleMessageErrorConstructor = createError(
  'UNEXPECTED_NOT_FOUND',
  '%s not found',
  500,
);

export class UnimplementedError extends Error {
  constructor(msg: string) {
    super('TODO: ' + msg);
  }
}

export class UnreachableError extends Error {}
