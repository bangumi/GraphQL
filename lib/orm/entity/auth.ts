import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';

@Index('access_token', ['accessToken'], { unique: true })
@Index('type', ['type'], {})
@Entity('chii_oauth_access_tokens', { schema: 'bangumi' })
export class OauthAccessTokens {
  @PrimaryGeneratedColumn({ type: 'mediumint', name: 'id' })
  id!: number;

  @Column('tinyint', { name: 'type', unsigned: true, default: () => "'0'" })
  type!: number;

  @Column('varchar', { name: 'access_token', unique: true, length: 40 })
  accessToken!: string;

  @Column('varchar', { name: 'client_id', length: 80 })
  clientId!: string;

  @Column('varchar', { name: 'user_id', nullable: true, length: 80 })
  userId!: string;

  @Column('timestamp', { name: 'expires', default: () => 'CURRENT_TIMESTAMP' })
  expires!: Date;

  @Column('varchar', { name: 'scope', nullable: true, length: 4000 })
  scope: string | undefined;

  @Column('varchar', { name: 'info', length: 255 })
  info!: string;
}

@Entity('chii_os_web_sessions', { schema: 'bangumi', name: 'chii_os_web_sessions' })
export class WebSessions {
  @Column('char', {
    primary: true,
    name: 'key',
    comment: 'session key',
    length: 64,
  })
  key!: string;

  @Column('int', { name: 'user_id', comment: 'uint32 user id', unsigned: true })
  userID!: number;

  @Column('mediumblob', { name: 'value', comment: 'json encoded session data' })
  value!: Buffer;

  @Column('bigint', {
    name: 'created_at',
    comment: 'int64 unix timestamp, when session is created',
  })
  createdAt!: number;

  @Column('bigint', {
    name: 'expired_at',
    comment: 'int64 unix timestamp, when session is expired',
  })
  expiredAt!: number;
}
