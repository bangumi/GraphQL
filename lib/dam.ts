/** 敏感词过滤 */
import RE2 from 're2';

import { fileConfig } from './config';

const controlPattern = /[\u200B-\u200F\u202A-\u202E\uFEFF]/;

interface Option {
  nsfw_word?: string;
  disabled_words?: string;
  banned_domain?: string;
}

export class Dam {
  private readonly nsfwWord: RE2 | undefined;
  private readonly disableWord: RE2 | undefined;
  private readonly bannedDomain: RE2 | undefined;

  constructor({ banned_domain, disabled_words, nsfw_word }: Option) {
    if (nsfw_word) {
      this.nsfwWord = new RE2(nsfw_word);
    }

    if (disabled_words) {
      this.disableWord = new RE2(disabled_words);
    }

    if (banned_domain) {
      this.bannedDomain = new RE2(banned_domain);
    }
  }

  needReview(text: string): boolean {
    if (text === '') {
      return false;
    }

    if (!this.disableWord) {
      return false;
    }

    return this.disableWord.test(text.toLowerCase());
  }

  censoredWords(text: string): boolean {
    if (this.disableWord?.test(text)) {
      return true;
    }

    if (this.bannedDomain != undefined) {
      return this.bannedDomain.test(text);
    }

    return false;
  }

  static allCharacterPrintable(s: string): boolean {
    return !controlPattern.test(s);
  }
}

export const dam = new Dam(fileConfig);
