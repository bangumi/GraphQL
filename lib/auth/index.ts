import { createError } from '@fastify/error';

import prisma from '../prisma';
import type { Permission } from './permission';
import { getPermission } from './permission';

export interface IUser {
  ID: number;
  username: string;
  nickname: string;
  img: string;
}

export interface IAuth {
  login: boolean;
  allowNsfw: boolean;
  permission: Readonly<Permission>;
  user: null | IUser;
}

export const TokenNotValidError = createError(
  'TOKEN_INVALID',
  "can't find user by access token",
  401,
);

export async function byHeader(key: string | string[] | undefined): Promise<IAuth> {
  if (Array.isArray(key)) {
    throw new HeaderInvalid("can't providing multiple access token");
  }
  if (!key) {
    return { login: false, permission: {}, allowNsfw: false, user: null };
  }
  if (!key.startsWith(tokenPrefix)) {
    throw new HeaderInvalid('authorization header should have "Bearer ${TOKEN}" format');
  }
  return await byToken(key.slice(tokenPrefix.length));
}

export async function byToken(access_token: string | undefined): Promise<IAuth> {
  if (!access_token) {
    return {
      user: null,
      login: false,
      permission: {},
      allowNsfw: false,
    };
  }
  const token = await prisma.chii_oauth_access_tokens.findFirst({
    where: { access_token, expires: { gte: new Date() } },
  });

  if (!token) {
    throw new TokenNotValidError();
  }

  const user = await prisma.chii_members.findFirst({
    where: { uid: Number.parseInt(token.user_id!) },
  });

  if (!user) {
    throw new Error(
      'missing user, please report a issue at https://github.com/bangumi/GraphQL/issues',
    );
  }

  return {
    user: {
      ID: user.uid,
      nickname: user.nickname,
      username: user.username,
      img: user.avatar,
    },
    login: true,
    permission: await getPermission(user.groupid),
    allowNsfw: user.regdate - Date.now() / 1000 <= 60 * 60 * 24 * 90,
  };
}

export const HeaderInvalid = createError('AUTHORIZATION_INVALID', '%s', 401);
export const tokenPrefix = 'Bearer ';
