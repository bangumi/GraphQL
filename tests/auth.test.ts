import { describe, test, expect } from 'vitest';

import * as auth from '@app/lib/auth';

describe('should auth', () => {
  test('current auth', async () => {
    const a = await auth.byToken('a_development_access_token');
    expect(a?.permission.report).toBe(true);

    const cached = await auth.byToken('a_development_access_token');
    expect(cached?.permission.report).toBe(true);
    expect(cached).toMatchObject({ groupID: 10 });
  });
});
