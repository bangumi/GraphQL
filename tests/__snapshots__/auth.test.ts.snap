// Vitest Snapshot v1

exports[`should auth > current auth 1`] = `
{
  "groupID": 10,
  "id": 382951,
  "img": "000/38/29/382951.jpg?r=1571167246",
  "nickname": "树洞酱",
  "regTime": 1512262276,
  "username": "382951",
}
`;

exports[`should auth > current auth 2`] = `
{
  "groupID": 10,
  "id": 382951,
  "img": "000/38/29/382951.jpg?r=1571167246",
  "nickname": "树洞酱",
  "regTime": 1512262276,
  "username": "382951",
}
`;
